File: typescript/day11/main.ts
import * as fs from "fs";

type Point = {
    x: number,
    y: number
}

const coordinates: (keyof Point)[] = ["x", "y"];

function load_data(filename: string): Point[] {
    let lines = fs.readFileSync(filename, 'utf-8').split("\n");
    let galaxies: Point[] = []
    lines.forEach((line, y) => {
        for (let x = 0; x < line.length; x++) {
            if (line[x] == '#') {
                galaxies.push({ x: x, y: y });
            }
        }
    });
    return galaxies;
}

function spread_galaxies(galaxies: Point[], spread_factor: number) {
    for (let coord of coordinates) {
        galaxies.sort((a, b) => a[coord] - b[coord]);
        let current_spread = 0;
        let current_pos = 0;
        for (let galaxy of galaxies) {
            let diff = galaxy[coord] - current_pos;
            current_pos = galaxy[coord];
            if (diff > 1) {
                current_spread += (diff - 1) * (spread_factor - 1);
            }
            galaxy[coord] += current_spread;
        }
    }
}

function sum_distance(galaxies: Point[]): number {
    let sum = 0;
    for (let i = 0; i < galaxies.length; i++) {
        for (let j = i + 1; j < galaxies.length; j++) {
            sum += Math.abs(galaxies[i].x - galaxies[j].x) + Math.abs(galaxies[i].y - galaxies[j].y);
        }
    }
    return sum;
}

function solution(filename: string, spread_factor: number): number {
    let galaxies = load_data(filename);
    spread_galaxies(galaxies, spread_factor);
    return sum_distance(galaxies);
}

function solution_part_1(filename: string) {
    console.log(solution(filename, 2));
}

function solution_part_2(filename: string) {
    console.log(solution(filename, 1000000));
}

solution_part_1("input.txt");
solution_part_2("input.txt");

File: Rust/day11/src/main.rs

#[derive(Debug)]
struct Point {
    x: usize,
    y: usize,
}

impl Point {
    fn distance(&self, other: &Point) -> usize {
        let x = self.x.abs_diff(other.x);
        let y = self.y.abs_diff(other.y);
        x + y
    }

}

fn load_data(filename: &str) -> Vec<Point> {
    let data = std::fs::read_to_string(filename).unwrap();
    let mut galaxies = Vec::new();
    for (y, line) in data.lines().enumerate() {
        for (x, c) in line.chars().enumerate() {
            if c == '#' {
                galaxies.push(Point { x, y });
            }
        }
    }
    galaxies
}

macro_rules! make_new_galaxy_positions {
    ($func_name:ident, $coord:ident) => {
        fn $func_name(galaxies: &mut [Point], spread_factor: usize) {
            galaxies.sort_by_key(|g| g.$coord);
            let mut current_spread = 0;
            let mut last_pos = 0;
            for galaxy in galaxies {
                if (galaxy.$coord - last_pos) > 1 {
                    current_spread += ((galaxy.$coord - last_pos) - 1) * (spread_factor - 1);
                }
                last_pos = galaxy.$coord;
                galaxy.$coord += current_spread;
            }
        }
    };
}

make_new_galaxy_positions!(calculate_new_galaxy_positions_x, x);
make_new_galaxy_positions!(calculate_new_galaxy_positions_y, y);

fn calculate_new_galaxy_positions(galaxies: &mut [Point], spread_factor: usize) {
    calculate_new_galaxy_positions_y(galaxies, spread_factor);
    calculate_new_galaxy_positions_x(galaxies, spread_factor);
}

#[allow(dead_code)]
fn print_map(map: &[Vec<bool>]) {
    for row in map {
        for column in row {
            if *column {
                print!("#");
            } else {
                print!(".");
            }
        }
        println!();
    }
}


fn sum_distances(galaxy: &Point, galaxies: &[Point]) -> usize {
    galaxies.iter().map(|g| galaxy.distance(g)).sum()
}

fn solution(spread_factor: usize) {
    let mut galaxies = load_data("input.txt");
    calculate_new_galaxy_positions(&mut galaxies, spread_factor);
    let total_sum = galaxies.iter().map(|g| sum_distances(g, &galaxies)).sum::<usize>();
    println!("Total sum: {}", total_sum / 2);
}

fn solution_part_1() {
    solution(2);
}

fn solution_part_2() {
    solution(1000000);
}

fn main() {
    solution_part_1();
    solution_part_2();
}
